source/backtraceMetric.ts

```diff
@@ -0,0 +1,226 @@
+    sessionId: string,
+    metricGroup: string,
+  ): Promise<void> {
+    const endpoint = `https://${this.DIRECTORY}.backtrace.io/api/summed-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    const payload = {
+      application: this.applicationName,
+      appversion: __VERSION__,
+      metadata: {
+        dropped_events: 0,
+      },
+      summed_events: [
+        {
+          timestamp: this.CURRENT_TIMESTAMP,
+          metric_group: metricGroup,
+          attributes: {
+            guid: sessionId,
+            'uname.sysname': this.userAgent,
+            'application.version': __VERSION__,
+            'application.session': '', // application session ?
+          },
+        },
+      ],
+    };
+
+    await this.post(endpoint, JSON.stringify(payload));
+    return;
+  }
+
+  /**
+   * Send POST request.
+   * @param url - string endpoint
+   * @param data - stringified JSON object
+   */
+  private async post(url: string, data: string): Promise<void> {
+    try {
+      return new Promise<void>((res, rej) => {
+        const http = new XMLHttpRequest();
+        http.setRequestHeader('Content-type', 'application/json');
+        http.timeout = this.timeout;
+        http.open('POST', url, true);
+        http.send(data);
+        http.onload = (e) => {
+          if (http.readyState === XMLHttpRequest.DONE) {
+            if (http.status === 200) {
+              res();
+            } else if (http.status === 429) {
+              rej('Backtrace - reached metric limit.');
+            } else {
+              rej(
+                `Invalid attempt to submit metric to Backtrace. Result: ${http.responseText}`,
+              );
+            }
+          }
+        };
+        http.onerror = (e) => {
+          rej(e);
+        };
+      });
+    } catch (err) {
+      console.error(err);
+    }
+  }
+
+  /**
+   * Get stored sessionId
+   */
+  private getSessionId(): string | undefined {
+    return localStorage.getItem('sessionId') || undefined;
+  }
+
+  /**
+   * Create new sessionId and set local sessionId and sessionStart.
+   */
+  private createNewSession(): string {
+    const newSessionId = uuid();
+    localStorage.setItem('sessionId', newSessionId);
+    localStorage.setItem('sessionStart', this.CURRENT_TIMESTAMP.toString());
+    this.sessionId = newSessionId;
+    return newSessionId;
+  }
+
+  /**
+   * Get stored time since last page navigation or current time
+   */
+  private getSessionStart(): number | undefined {
+    const sessionStartStr = localStorage.getItem('sessionStart');
+    return sessionStartStr ? Number(sessionStartStr) : undefined;
+  }
+
+  /**
+   * Get stored time since last session created or current time
+   */
+  private getLastActive(): number {
+    const lastActiveStr = localStorage.getItem('lastActive');
+    if (!lastActiveStr) {
+      this.setLastActive(this.CURRENT_TIMESTAMP);
+      return this.CURRENT_TIMESTAMP;
+    }
+    return Number(lastActiveStr);
```

Comment:
do we need to pass it to the `Number` constructor? I believe it's dangerous. Shouldn't  we just try to parse a number instead?

Reply:
Changing to parseInt